text/0000-type-ascription.md

```diff
@@ -0,0 +1,175 @@
+- Start Date: 2015-2-3
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add type ascription to expressions and patterns.
+
+Type ascription on expression has already been implemented. Type ascription on
+patterns can probably wait until post-1.0.
+
+See also discussion on [#354](https://github.com/rust-lang/rfcs/issues/354) and
+[rust issue 10502](https://github.com/rust-lang/rust/issues/10502).
+
+
+# Motivation
+
+Type inference is imperfect. It is often useful to help type inference by
+annotating a sub-expression or sub-pattern with a type. Currently, this is only
+possible by extracting the sub-expression into a variable using a `let`
+statement and/or giving a type for a whole expression or pattern. This is un-
+ergonomic, and sometimes impossible due to lifetime issues. Specifically, a
+variable has lifetime of its enclosing scope, but a sub-expression's lifetime is
+typically limited to the nearest semi-colon.
+
+Typical use cases are where a function's return type is generic (e.g., collect)
+and where we want to force a coercion.
+
+Type ascription can also be used for documentation and debugging - where it is
+unclear from the code which type will be inferred, type ascription can be used
+to precisely communicate expectations to the compiler or other programmers.
+
+By allowing type ascription in more places, we remove the inconsistency that
+type ascription is currently only allowed on top-level patterns.
+
+## Examples:
+
+Generic return type:
+
+```
+// Current.
+let z = if ... {
+    let x: Vec<_> = foo.enumerate().collect();
+    x
+} else {
+    ...
+};
+
+// With type ascription.
+let z = if ... {
+    foo.enumerate().collect(): Vec<_>
+} else {
+    ...
+};
+```
+
+Coercion:
+
+```
+fn foo<T>(a: T, b: T) { ... }
+
+// Current.
+let x = [1u32, 2, 4];
+let y = [3u32];
+...
+let x: &[_] = &x;
+let y: &[_] = &y;
+foo(x, y);
+
+// With type ascription.
+let x = [1u32, 2, 4];
+let y = [3u32];
+...
+foo(x: &[_], y: &[_]);
+```
+
+In patterns:
+
+```
+struct Foo<T> { a: T, b: String }
+
+// Current
+fn foo(Foo { a, .. }: Foo<i32>) { ... }
+
+// With type ascription.
+fn foo(Foo { a: i32, .. }) { ... }
+```
```

Comment:
Good question. I don't think there is a good answer. I think the least worst is to assume that users will prefer the common pattern of using the same name for both the name of the field and the new variable (`fn foo(Foo { a, .. }: ...` in the current syntax), then assume that a single `:` in a pattern always denotes a type, i.e., we assume `Bar` is always a type (this is backwards incompatible, as you allude to). If a user wants to rename the variable, then they'd have to use `a : b : _`, which is bad. I think the alternative is theoretically nicer - type ascription should be more optional, but less practical, since it is common to reuse the field name.